File: source/downloads/code/LeetCode/Python/Swap-Nodes-in-Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
            return head

        front = head
        end = front.next
        prev = ListNode(-1)
        prev.next = head
        newhead = head.next
        
        while end!=None:
            nextNode = end.next
            front.next = nextNode
            end.next = front
            prev.next = end
            
            if nextNode == None:
                break
            prev = front
            front = nextNode
            end = front.next

        return newhead

File: downloads/code/LeetCode/Java/Swap-Nodes-in-Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        prev.next = head;
        if(head==null || head.next==null)
            return head;
        
        while (prev.next!=null && prev.next.next!=null){
            ListNode first = prev.next;
            ListNode second = prev.next.next;
    
            first.next = second.next;        
            prev.next = second;
            prev.next.next = first;
            
            prev = prev.next.next;
            
        }
        return dummy.next;
    }
}
